rfcs/proposed/numa_support/create-numa-arenas.md

```diff
@@ -0,0 +1,224 @@
+  std::vector<tbb::task_group> task_groups(numa_nodes.size());
+
+  // initialize each arena, each constrained to a different NUMA node
+  for (int i = 0; i < numa_nodes.size(); i++)
+    arenas[i].initialize(tbb::task_arena::constraints(numa_nodes[i]), 0);
+```
+
+The first line obtains a vector of NUMA node IDs for the system. Then, a vector of the same size
+is created to store `tbb::task_arena` objects, each constrained to one of the NUMA nodes.
+Another vector holds `task_group` instances used later to submit and wait for completion
+of the work in each of the arenas - it is necessary because `task_arena` does not provide
+any work synchronization API. Finally, the loop over all NUMA nodes initializes associated
+task arenas with proper constraints.
+
+While not incomprehensible, the code is quite verbose and arguably too explicit for the typical scenario
+of creating a set of arenas across all available NUMA domains. There is also risk of subtle issues.
+The default constructor of `task_arena` reserves a slot for an application thread. The arena initialization
+at the last line explicitly overwrites it to 0 to allow TBB worker threads taking all the slots, however
+this nuance might be unknown and easy to miss, potentially resulting in underutilization of CPU resources.
+
+## Proposal
+
+We propose to introduce a special function to create the set of task arenas, one per NUMA node on the system.
+The initialization code equivalent to the example above would be:
+
+```c++
+  std::vector<tbb::task_arena> arenas = tbb::create_numa_task_arenas();
+  std::vector<tbb::task_group> task_groups(arenas.size());
+```
+
+The rest of the code in that example might be rewritten with the API proposed in
+[Waiting in a task arena](../task_arena_waiting/readme.md):
+
+```c++
+  // enqueue work to all but the first arena, using the task groups to track work
+  for (int i = 1; i < arenas.size(); i++)
+    arenas[i].enqueue(
+      [] { tbb::parallel_for(0, N, [](int j) { f(w); }); }, 
+      task_groups[i]
+    );
+
+  // directly execute the work to completion in the remaining arena
+  arenas[0].execute([] {
+    tbb::parallel_for(0, N, [](int j) { f(w); });
+  });
+
+  // join the other arenas to wait on their task groups
+  for (int i = 1; i < arenas.size(); i++)
+    arenas[i].wait_for(task_groups[i]);
+```
+
+### Public API
+
+The function has the following signature:
+
+```c++
+// Defined in tbb/task_arena.h
+
+namespace tbb {
+    std::vector<tbb::task_arena> create_numa_task_arenas(
+        task_arena::constraints other_constraints = {},
+        unsigned reserved_slots = 0
+    };
+}
+```
+
+It optionally takes a `constraints` argument to change default arena settings such as maximal concurrency
+(the upper limit on the number of threads), core type etc.; the `numa_id` value in `other_constraints`
+is ignored. The second optional argument allows to override the number of reserved slots, which by default
```

Comment:
Would it be more practical to raise the error if the user provided the `constraints` with some `numa_id` specified instead of ignoring it? For me, it seems like the user is doing something wrong in this case and it would be good to warn the user somehow.

Reply:
Then there should be some "valid" value. What would that be? `task_arena::automatic`?

If not just ignoring other values, then I would go for undefined behavior, which would mean "ignore in release builds, assert in debug builds".

Reply:
I have added an open for that.,